File: src/main/java/io/protobase/f7/formulas/math/POW.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.errors.NumException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class POW extends AbstractFormula {
  public static FormulaName NAME = FormulaName.POW;
  public static POW SELF = new POW();

  public POW() {
    super();
  }

  public POW(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 2, NAME);
    Object first = Converters.first(collateralLookup.apply(origin, values[0]));
    Object second = Converters.first(collateralLookup.apply(origin, values[1]));
    if (first instanceof F7Exception) {
      return first;
    }
    if (second instanceof F7Exception) {
      return second;
    }
    double result = Math.pow(Converters.toDouble(first), Converters.toDouble(second));
    if (Double.isNaN(result) || Double.isInfinite(result)) {
      return new NumException("POW evaluates to an imaginary number.");
    }
    return result;
  }
}

File: src/main/js/formulas/math/POW.ts
import { F7Exception } from "../../errors/F7Exception";
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class POW extends AbstractFormula {
  static SELF: POW = new POW();
  NAME = FormulaName.POW;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 2, this.NAME);
    const first = Converters.first(this.collateralLookup(origin, values[0]));
    const second = Converters.first(this.collateralLookup(origin, values[1]));
    if (first instanceof F7Exception) {
      return first;
    }
    if (second instanceof F7Exception) {
      return second;
    }
    const firstNumber = Converters.toNumber(first);
    const secondNumber = Converters.toNumber(second);
    const result = Math.pow(firstNumber, secondNumber);
    if (isNaN(result) || !Number.isFinite(result)) {
      return new NumException("POW evaluates to an imaginary number.");
    }
    return result;
  }
}
